pkg/render/intrusiondetection/dpi/dpi.go

```diff
@@ -0,0 +1,391 @@
+	"github.com/tigera/operator/pkg/components"
+	"github.com/tigera/operator/pkg/ptr"
+	"github.com/tigera/operator/pkg/render"
+	"github.com/tigera/operator/pkg/render/common/configmap"
+	"github.com/tigera/operator/pkg/render/common/meta"
+	rmeta "github.com/tigera/operator/pkg/render/common/meta"
+	"github.com/tigera/operator/pkg/render/common/secret"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	DeepPacketInspectionNamespace = "tigera-dpi"
+	DeepPacketInspectionName      = "tigera-dpi"
+	DefaultMemoryLimitDPI         = "1Gi"
+	DefaultMemoryRequestDPI       = "100Mi"
+	DefaultCPULimitDPI            = "1"
+	DefaultCPURequestDPI          = "100m"
+)
+
+func DPI(
+	ids *operatorv1.IntrusionDetection,
+	installation *operatorv1.InstallationSpec,
+	nodeTLSSecret *corev1.Secret,
+	typhaTLSSecret *corev1.Secret,
+	typhaCAConfigMap *corev1.ConfigMap,
+	pullSecrets []*corev1.Secret,
+	openshift bool,
+	hasNoLicense bool,
+) render.Component {
+	return &dpi{
+		ids:              ids,
+		installation:     installation,
+		nodeTLSSecret:    nodeTLSSecret,
+		typhaTLSSecret:   typhaTLSSecret,
+		typhaCAConfigMap: typhaCAConfigMap,
+		pullSecrets:      pullSecrets,
+		openshift:        openshift,
+		hasNoLicense:     hasNoLicense,
+	}
+}
+
+type dpi struct {
+	ids              *operatorv1.IntrusionDetection
+	installation     *operatorv1.InstallationSpec
+	nodeTLSSecret    *corev1.Secret
+	typhaTLSSecret   *corev1.Secret
+	typhaCAConfigMap *corev1.ConfigMap
+	pullSecrets      []*corev1.Secret
+	openshift        bool
+	hasNoLicense     bool
+	dpiImage         string
+}
+
+func (d *dpi) ResolveImages(is *operatorv1.ImageSet) error {
+	reg := d.installation.Registry
+	path := d.installation.ImagePath
+	prefix := d.installation.ImagePrefix
+	var err error
+	d.dpiImage, err = components.GetReference(components.ComponentDeepPacketInspection, reg, path, prefix, is)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *dpi) Objects() (objsToCreate, objsToDelete []client.Object) {
+	var toDelete []client.Object
+	var toCreate []client.Object
+	// Add DeepPacketInspection related objects if all necessary bit of TLS config are set.
+	addDPIDaemonSet := d.nodeTLSSecret != nil && d.typhaTLSSecret != nil && d.typhaCAConfigMap != nil
+
+	// toCreate is populated only if DPI daemonset needs to be created and there is valid license,
+	// for all other combinations delete the DPI namespace.
+	if addDPIDaemonSet && !d.hasNoLicense {
+		// Update the IDS resource with the default values for DPI ComponentResources if it doesn't exist,
+		// these values are used for setting container resource requirements in DPI daemonset.
+		d.defaultIntrusionDetectionComponentResources()
+		toCreate = append(toCreate, d.ids)
```

Comment:
I think you probably pulled this from logstorage, but I don't think this is the correct way to write back the defaults, even more so because this could be done by the intrusion detection render code.

If this is returned as an object from the render code the owner reference will be changed.

Reply:
Okay, so I guess the default values should be set by intrusion detection controller before objects returned by intrusion detection render are created?

Reply:
Yes exactly, set the default before anything is rendered and write them back. I think the core controller has a good example of this.